### ISSUE TITLE
Logging level remains captured by SpringBoot after unittests complete

### ISSUE THREAD

**Issue by dev930018** [spring-projects/spring-boot#19323]
After running a `SpringBootTest` with a `SpringRunner`, there continue to be effects to Loggers. In particular, `DEBUG` log messages are NO-OPed out but `INFO` log messages are unchanged.

### `pom.xml`:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>spring-bug</groupId>
    <artifactId>minimal-example</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <spring.version>2.1.9.RELEASE</spring.version>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot</artifactId>
            <version>${spring.version}</version>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter</artifactId>
            <version>${spring.version}</version>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <version>${spring.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-test</artifactId>
            <scope>test</scope>
            <version>${spring.version}</version>
        </dependency>
    </dependencies>

</project>
```

### `ASpringRunnerTest.java`:
```
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringRunner;

import static org.junit.Assert.assertTrue;

@SpringBootTest
@RunWith(SpringRunner.class)
@ContextConfiguration(classes = SpringBeans.class)
@ActiveProfiles(profiles = "")
public class ASpringRunnerTest {

    @Autowired
    String someBean;

    @Test
    public void testWithSpringRunner() {
        assertTrue(someBean.equals(new SpringBeans().stringFactory()));
    }

}
```

### `BJunitLoggingTest.java`:
```
import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.core.read.ListAppender;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.slf4j.LoggerFactory;

import java.util.Collections;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static org.junit.Assert.assertEquals;

@RunWith(JUnit4.class)
public class BJunitLoggingTest {

    private Logger logger;
    private ListAppender<ILoggingEvent> appender;

    @Before
    public void setUp() {
        logger = (Logger) LoggerFactory.getLogger(BJunitLoggingTest.class);
        appender = new ListAppender<>();
        appender.start();
        logger.addAppender(appender);
    }

    @After
    public void tearDown() {
        logger.detachAppender(appender);
        appender.stop();
    }

    private List<String> getMessages(Predicate<ILoggingEvent> predicate) {
        return appender.list.stream()
                .filter(predicate)
                .map(ILoggingEvent::getFormattedMessage)
                .collect(Collectors.toList());
    }

    @Test
    public void debugTest() {
        // Given
        logger.debug("test string");
        // When
        List<String> infoMessages = getMessages(event -> event.getLevel() == Level.DEBUG);
        // Then
        assertEquals(infoMessages, Collections.singletonList("test string"));
    }


    @Test
    public void infoTest() {
        // Given
        logger.info("test string");
        // When
        List<String> infoMessages = getMessages(event -> event.getLevel() == Level.INFO);
        // Then
        assertEquals(infoMessages, Collections.singletonList("test string"));
    }

}
```

### `SpringBeans.java`:
```
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class SpringBeans {

    @Bean
    public String stringFactory() {
        return "string factory";
    }

}
```

### `mvn test -Dtest=BJunitLoggingTest`:
This skips any instantiation of SpringBoot stuff
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running BJunitLoggingTest
13:27:47.837 [main] DEBUG BJunitLoggingTest - test string
13:27:47.853 [main] INFO BJunitLoggingTest - test string
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.284 sec

Results :

Tests run: 2, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.569 s
[INFO] Finished at: 2019-12-06T13:27:47Z
[INFO] ------------------------------------------------------------------------
```

### `mvn test`:
```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.1.9.RELEASE)

2019-12-06 13:29:52.128  INFO 13422 --- [           main] ASpringRunnerTest                        : Starting ASpringRunnerTest on XXX with PID 13422 (started by XXX in /mnt/c/dev/minimalexample)
2019-12-06 13:29:52.130  INFO 13422 --- [           main] ASpringRunnerTest                        : No active profile set, falling back to default profiles: default
2019-12-06 13:29:52.504  INFO 13422 --- [           main] ASpringRunnerTest                        : Started ASpringRunnerTest in 0.76 seconds (JVM running for 1.726)
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.879 sec
Running BJunitLoggingTest
2019-12-06 13:29:52.962  INFO 13422 --- [           main] BJunitLoggingTest                        : test string
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec <<< FAILURE!
debugTest(BJunitLoggingTest)  Time elapsed: 0.002 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[]> but was:<[test string]>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at BJunitLoggingTest.debugTest(BJunitLoggingTest.java:53)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)


Results :

Failed tests:   debugTest(BJunitLoggingTest): expected:<[]> but was:<[test string]>

Tests run: 3, Failures: 1, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.765 s
[INFO] Finished at: 2019-12-06T13:29:53Z
[INFO] ------------------------------------------------------------------------
```

**Comment by dev930018** [comment]
## Current potential workaround
`test/resources/application.yaml`:
```
logging:
  level:
    root: DEBUG
```

`mvn test`:
```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.1.9.RELEASE)

2019-12-06 13:40:28.748  INFO 13580 --- [           main] ASpringRunnerTest                        : Starting ASpringRunnerTest on ESD-LTGSL23X2 with PID 13580 (started by lassiter_at in /mnt/c/dev/minimalexample)
2019-12-06 13:40:28.749 DEBUG 13580 --- [           main] ASpringRunnerTest                        : Running with Spring Boot v2.1.9.RELEASE, Spring v5.1.10.RELEASE
2019-12-06 13:40:28.749  INFO 13580 --- [           main] ASpringRunnerTest                        : No active profile set, falling back to default profiles: default
2019-12-06 13:40:28.749 DEBUG 13580 --- [           main] o.s.boot.SpringApplication               : Loading source class SpringBeans
2019-12-06 13:40:28.778 DEBUG 13580 --- [           main] o.s.b.c.c.ConfigFileApplicationListener  : Loaded config file 'file:/mnt/c/dev/minimalexample/target/test-classes/application.yaml' (classpath:/application.yaml)
2019-12-06 13:40:28.779 DEBUG 13580 --- [           main] s.c.a.AnnotationConfigApplicationContext : Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@435fb7b5
2019-12-06 13:40:28.800 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'org.springframework.context.annotation.internalConfigurationAnnotationProcessor'
2019-12-06 13:40:28.830 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'org.springframework.boot.autoconfigure.internalCachingMetadataReaderFactory'
2019-12-06 13:40:29.190 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'org.springframework.boot.test.mock.mockito.MockitoPostProcessor'
2019-12-06 13:40:29.251 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'org.springframework.context.event.internalEventListenerProcessor'
2019-12-06 13:40:29.255 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'org.springframework.context.event.internalEventListenerFactory'
2019-12-06 13:40:29.267 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'org.springframework.context.annotation.internalAutowiredAnnotationProcessor'
2019-12-06 13:40:29.270 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'org.springframework.context.annotation.internalCommonAnnotationProcessor'
2019-12-06 13:40:29.284 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'org.springframework.boot.test.mock.mockito.MockitoPostProcessor$SpyPostProcessor'
2019-12-06 13:40:29.296 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'springBeans'
2019-12-06 13:40:29.319 DEBUG 13580 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Creating shared instance of singleton bean 'stringFactory'
2019-12-06 13:40:29.402  INFO 13580 --- [           main] ASpringRunnerTest                        : Started ASpringRunnerTest in 1.166 seconds (JVM running for 2.327)
2019-12-06 13:40:29.407 DEBUG 13580 --- [           main] c.DefaultCacheAwareContextLoaderDelegate : Storing ApplicationContext in cache under key [[MergedContextConfiguration@2a693f59 testClass = ASpringRunnerTest, locations = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', p
ropertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework.bo
ot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@7eac9008, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], contextLoa
der = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]]]
2019-12-06 13:40:29.408 DEBUG 13580 --- [           main] org.springframework.test.context.cache   : Spring test ApplicationContext cache statistics: [DefaultContextCache@6c44052e size = 1, maxSize = 32, parentContextCount = 0, hitCount = 0, missCount = 1]
2019-12-06 13:40:29.431 DEBUG 13580 --- [           main] tractDirtiesContextTestExecutionListener : Before test method: context [DefaultTestContext@433defed testClass = ASpringRunnerTest, testInstance = ASpringRunnerTest@5965d37, testMethod = testWithSpringRunner@ASpringRunnerTest, testException = [null], mergedContextConfiguration = [MergedContextConfig
uration@2a693f59 testClass = ASpringRunnerTest, locations = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoc
onfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContex
tCustomizer@7eac9008, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map[[empty]]], class annotated with @DirtiesContext [false] with m
ode [null], method annotated with @DirtiesContext [false] with mode [null].
2019-12-06 13:40:29.433 DEBUG 13580 --- [           main] c.DefaultCacheAwareContextLoaderDelegate : Retrieved ApplicationContext from cache with key [[MergedContextConfiguration@2a693f59 testClass = ASpringRunnerTest, locations = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}'
, propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework
.boot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@7eac9008, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], context
Loader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]]]
2019-12-06 13:40:29.434 DEBUG 13580 --- [           main] org.springframework.test.context.cache   : Spring test ApplicationContext cache statistics: [DefaultContextCache@6c44052e size = 1, maxSize = 32, parentContextCount = 0, hitCount = 1, missCount = 1]
2019-12-06 13:40:29.934 DEBUG 13580 --- [           main] c.DefaultCacheAwareContextLoaderDelegate : Retrieved ApplicationContext from cache with key [[MergedContextConfiguration@2a693f59 testClass = ASpringRunnerTest, locations = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}'
, propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework
.boot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@7eac9008, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], context
Loader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]]]
2019-12-06 13:40:29.935 DEBUG 13580 --- [           main] org.springframework.test.context.cache   : Spring test ApplicationContext cache statistics: [DefaultContextCache@6c44052e size = 1, maxSize = 32, parentContextCount = 0, hitCount = 2, missCount = 1]
2019-12-06 13:40:29.935 DEBUG 13580 --- [           main] c.DefaultCacheAwareContextLoaderDelegate : Retrieved ApplicationContext from cache with key [[MergedContextConfiguration@2a693f59 testClass = ASpringRunnerTest, locations = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}'
, propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework
.boot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@7eac9008, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], context
Loader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]]]
2019-12-06 13:40:29.936 DEBUG 13580 --- [           main] org.springframework.test.context.cache   : Spring test ApplicationContext cache statistics: [DefaultContextCache@6c44052e size = 1, maxSize = 32, parentContextCount = 0, hitCount = 3, missCount = 1]
2019-12-06 13:40:29.936 DEBUG 13580 --- [           main] c.DefaultCacheAwareContextLoaderDelegate : Retrieved ApplicationContext from cache with key [[MergedContextConfiguration@2a693f59 testClass = ASpringRunnerTest, locations = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}'
, propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework
.boot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@7eac9008, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], context
Loader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]]]
2019-12-06 13:40:29.936 DEBUG 13580 --- [           main] org.springframework.test.context.cache   : Spring test ApplicationContext cache statistics: [DefaultContextCache@6c44052e size = 1, maxSize = 32, parentContextCount = 0, hitCount = 4, missCount = 1]
2019-12-06 13:40:29.939 DEBUG 13580 --- [           main] c.DefaultCacheAwareContextLoaderDelegate : Retrieved ApplicationContext from cache with key [[MergedContextConfiguration@2a693f59 testClass = ASpringRunnerTest, locations = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}'
, propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework
.boot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@7eac9008, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], context
Loader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]]]
2019-12-06 13:40:29.939 DEBUG 13580 --- [           main] org.springframework.test.context.cache   : Spring test ApplicationContext cache statistics: [DefaultContextCache@6c44052e size = 1, maxSize = 32, parentContextCount = 0, hitCount = 5, missCount = 1]
2019-12-06 13:40:29.942 DEBUG 13580 --- [           main] tractDirtiesContextTestExecutionListener : After test method: context [DefaultTestContext@433defed testClass = ASpringRunnerTest, testInstance = ASpringRunnerTest@5965d37, testMethod = testWithSpringRunner@ASpringRunnerTest, testException = [null], mergedContextConfiguration = [MergedContextConfigu
ration@2a693f59 testClass = ASpringRunnerTest, locations = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoco
nfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContext
Customizer@7eac9008, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map[[empty]]], class annotated with @DirtiesContext [false] with mo
de [null], method annotated with @DirtiesContext [false] with mode [null].
2019-12-06 13:40:29.944 DEBUG 13580 --- [           main] tractDirtiesContextTestExecutionListener : After test class: context [DefaultTestContext@433defed testClass = ASpringRunnerTest, testInstance = [null], testMethod = [null], testException = [null], mergedContextConfiguration = [MergedContextConfiguration@2a693f59 testClass = ASpringRunnerTest, locat
ions = '{}', classes = '{class SpringBeans}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@
0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@5bda8e08, org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@3b0090a4, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@7eac9008, org.springframework.boot.test.m
ock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@5dda768f], contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map[[empty]]], class annotated with @DirtiesContext [false] with mode [null].
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.475 sec
Running BJunitLoggingTest
2019-12-06 13:40:29.952 DEBUG 13580 --- [           main] BJunitLoggingTest                        : test string
2019-12-06 13:40:29.954  INFO 13580 --- [           main] BJunitLoggingTest                        : test string
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
2019-12-06 13:40:29.971 DEBUG 13580 --- [       Thread-0] s.c.a.AnnotationConfigApplicationContext : Closing org.springframework.context.annotation.AnnotationConfigApplicationContext@435fb7b5, started on Fri Dec 06 13:40:28 GMT 2019

Results :

Tests run: 3, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.478 s
[INFO] Finished at: 2019-12-06T13:40:30Z
[INFO] ------------------------------------------------------------------------
```

**Comment by philwebb** [comment]
@dev930018 Are you able to share those sample snippets as a complete project on GitHub? It will save us having to reconstruct them into a full project ourselves.

**Comment by dev930018** [comment]
[minimalexample.zip](https://github.com/spring-projects/spring-boot/files/3935149/minimalexample.zip)

**Comment by wilkinsona** [comment]
Thanks for the sample.

The root cause of the problem is that, like all logging framework's, the lifecycle of Logback's configuration is scoped to the `ClassLoader` that loaded it. When `ASpringRunnerTest` runs, the application context is refreshed and Spring Boot configures Logback as it would when starting the application at runtime. The test framework then caches this context in case it can be reused in the same test suite. This means that there's no opportunity for Logback to be unconfigured before `BJunitLoggingTest` runs.

Marking `ASpringRunnerTest` with `@DirtiesContext` would prevent the context from being cached such that it's closed after `ASpringRunnerTest` has completed. This would provide an opportunity for Logback to be unconfigured but I'm not sure that it's one that is worth taking. Other tests that do not use `@DirtiesContext` may reintroduce the problem. You'd also encounter problems if you ever wanted to run multiple tests in parallel.

In short, tests that make assertions about logging output will always be brittle. For any such test, I think your best option is to use a `@Before` method to ensure that the logging framework is configured as required by the test, including log levels, before the test runs. Ensuring that the logging framework is in a known consistent state before each test executes will ensure that you do not see tests pass or fail when they should not do so. We do this ourselves in a few places in the `spring-boot` module's tests. It does not prevent problems with parallel execution of tests in the same JVM, but that's something that we're happy to live with.
